File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example/h10-w/low_level/bms_state.cpp
// bms_state.cpp

#include <iostream>
#include <thread>
#include <chrono>

#include "common/common_def.h"
#include "common/h10w_def.h"
#include "h10-w/h10w_low_level.h"

using namespace humanoid;
using namespace humanoid::h10w;

int main() {
  std::cout << "=== BMS State Reading Test ===" << std::endl;

  H10wLowLevel low_level;
  BmsState bms_state;

  // Read BMS state for 10 seconds
  constexpr int DURATION_MS = 10000;
  constexpr int READ_INTERVAL_MS = 1000;
  constexpr int read_count = DURATION_MS / READ_INTERVAL_MS;

  for (int i = 0; i < read_count; ++i) {
    auto start = std::chrono::high_resolution_clock::now();

    // Read BMS state
    int ret = low_level.readBmsState(bms_state);
    if (ret != 0) {
      std::cout << "Failed to read BMS state!" << std::endl;
      return -1;
    }

    // Print BMS state information
    std::cout << "=== BMS State Reading " << i + 1 << " ===" << std::endl;
    std::cout << "Power Mode: " << static_cast<int>(bms_state.power_mode)
              << (bms_state.power_mode == 0 ? " (External)" : " (Battery)")
              << std::endl;
    std::cout << "Charge State: " << static_cast<int>(bms_state.charge_state);
    switch (bms_state.charge_state) {
      case 0:
        std::cout << " (Invalid)" << std::endl;
        break;
      case 1:
        std::cout << " (Charging)" << std::endl;
        break;
      case 2:
        std::cout << " (Discharging)" << std::endl;
        break;
      case 3:
        std::cout << " (Idle)" << std::endl;
        break;
      default:
        std::cout << " (Unknown)" << std::endl;
        break;
    }
    std::cout << "SOC: " << bms_state.soc << "%" << std::endl;
    std::cout << std::endl;

    auto end = std::chrono::high_resolution_clock::now();
    auto elapsed =
        std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
    std::this_thread::sleep_for(std::chrono::milliseconds(READ_INTERVAL_MS) -
                                elapsed);
  }

  std::cout << "BMS state reading test completed successfully" << std::endl;
  return 0;
}

File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example_py/h10-w/low_level/bms_state.py
#!/usr/bin/env python3

import humanoid_sdk_py
from humanoid_sdk_py import h10w
import time


def main():
    print("=== BMS State Reading Test ===")

    # Initialize low level object
    low_level = h10w.H10wLowLevel()

    # Read BMS state for 10 seconds
    DURATION_MS = 10000
    READ_INTERVAL_MS = 1000
    read_count = DURATION_MS // READ_INTERVAL_MS

    for i in range(read_count):
        start_time = time.time()

        # Read BMS state
        ret, bms_state = low_level.readBmsState()
        if ret != 0:
            print("Failed to read BMS state!")
            return -1

        # Print BMS state information
        print(f"=== BMS State Reading {i + 1} ===")
        print(f"Power Mode: {bms_state.power_mode} ({'External' if bms_state.power_mode == 0 else 'Battery'})")
        charge_states = {0: "Invalid", 1: "Charging", 2: "Discharging", 3: "Idle"}
        print(f"Charge State: {bms_state.charge_state} ({charge_states.get(bms_state.charge_state, 'Unknown')})")
        print(f"SOC: {bms_state.soc:.1f}%")
        print()

        elapsed_time = (time.time() - start_time) * 1000
        sleep_time = READ_INTERVAL_MS - elapsed_time
        if sleep_time > 0:
            time.sleep(sleep_time / 1000.0)

    print("BMS state reading test completed successfully")
    return 0


if __name__ == "__main__":
    exit(main())
